Harmony/Tools/AccessTools.cs

```diff
@@ -784,161 +782,520 @@ public static object[] ActualParameters(MethodBase method, object[] inputs)
+			if (instance is null)
+				throw new ArgumentNullException(nameof(instance));
+			if (fieldName is null)
+				throw new ArgumentNullException(nameof(fieldName));
+			try
+			{
+				return ref FieldRefAccessInternal<T, F>(GetInstanceField(typeof(T), fieldName), needCastclass: false)(instance);
+			}
+			catch (Exception ex)
+			{
+				throw new ArgumentException($"FieldRefAccess<{typeof(T)}, {typeof(F)}> for {instance}, {fieldName} caused an exception", ex);
+			}
 		}
 
-		/// <summary>Creates an instance field reference delegate for a private type</summary>
-		/// <typeparam name="F">The type of the field</typeparam>
-		/// <param name="type">The class/type</param>
+		/// <summary>Creates a field reference delegate for an instance field of a class or static field (NOT an instance field of a struct)</summary>
+		/// <typeparam name="F">
+		/// The type of the field; or if the field's type is a reference type (a class or interface, NOT a struct or other value type),
+		/// a type that <see cref="Type.IsAssignableFrom(Type)">is assignable from</see> the field's type
+		/// </typeparam>
+		/// <param name="type">
+		/// The type that defines the field, or derived class of this type; must not be a struct type unless the field is static
+		/// </param>
 		/// <param name="fieldName">The name of the field</param>
-		/// <returns>A read and writable <see cref="FieldRef{T,F}"/> delegate</returns>
+		/// <returns>
+		/// A readable/assignable <see cref="FieldRef{T,F}"/> delegate with <c>T=object</c>
+		/// (for static fields, the <c>instance</c> delegate parameter is ignored)
+		/// </returns>
+		/// <remarks>
+		/// This method supports static fields, even those defined in structs, for legacy reasons.
+		/// Consider using <see cref="StaticFieldRefAccess{F}(Type, string)"/> (and other overloads) instead for static fields.
+		/// </remarks>
 		///
 		public static FieldRef<object, F> FieldRefAccess<F>(Type type, string fieldName)
 		{
-			return FieldRefAccess<object, F>(Field(type, fieldName));
+			if (type is null)
+				throw new ArgumentNullException(nameof(type));
+			if (fieldName is null)
+				throw new ArgumentNullException(nameof(fieldName));
+			try
+			{
+				var fieldInfo = Field(type, fieldName);
+				if (fieldInfo is null)
+					throw new MissingFieldException(type.Name, fieldName);
+				// Backwards compatibility: This supports static fields, even those defined in structs. For static fields, T is effectively ignored.
+				if (fieldInfo.IsStatic is false && fieldInfo.DeclaringType is Type declaringType)
+				{
+					// When fieldInfo is passed to FieldRefAccess methods, the T generic class constraint is insufficient to ensure that
+					// the field is not a struct instance field, since T could be object, ValueType, or an interface that the struct implements.
+					if (declaringType.IsValueType)
+						throw new ArgumentException("Either FieldDeclaringType must be a class or field must be static");
```

Comment:
Arguably yes, since the alternative `MissingFieldException` indicates that a field can't be found, not that the field is invalid (for the method). In fact, it's arguable that `MissingFieldException` should be never be thrown from non-system code, given its description: https://docs.microsoft.com/en-us/dotnet/api/system.missingfieldexception